torch/csrc/jit/passes/bailout_graph.cpp

```diff
@@ -125,11 +137,53 @@ struct BailOutGraphBuilderForNode {
     updated_max_trip_count =
         copy_graph_->insert(aten::sub, {updated_max_trip_count, one});
     TORCH_INTERNAL_ASSERT(old_to_new_.count(outer_node->inputs()[0]) != 0);
-    mapValueAndCopyMetadata(outer_node->inputs()[0], updated_max_trip_count);
     auto cur_plus_one = copy_graph_->insert(aten::add, {one, cur_iter});
 
+    // We need to be careful when mapping `block_outputs` to continuation
+    // loop's inputs since `cloneFrom` will replace `%4` with the same value
+    // in both, `prim::Loop` and `aten::cat` in the example below:
+    //
+    // ... : Tensor = prim::Loop(%MAX_TRIP_COUNT, %COND, ..., %4)
+    //   block0(%i.2 : int, ...):
+    //     ...
+    //     %y.5 : Double(3) = aten::cat(%22, %4)
+    //     ...
+    //
+    // However for the cloned loop node, the values should be different.
+    // Namely, the value in `prim::Loop` should come from
+    // `lv.bodyBlock()->outputs()` which are mapped to the outputs of the
+    // current iteration whereas `%4` in `aten::cat` needs to be mapped to the
+    // cloned value of `%4` in a bailout graph.
+
+    // remember the continuation loop carried dependencies
+    // these will become `prim::Loop` inputs after cloning
+    auto continuation_block_outputs = fmap(
+        block_outputs, [this](Value* v) { return getOrAddInputForValue(v); });
+
+    // add a dummy input for every loop input that we haven't seen
```

Comment:
I am having difficult following the intent here. getOrAddInputForValue seems like it has a very clear meaning: get me this value or make an input for it and return me it in the new graph. In the new graph, however, the input to Loop is no longer the inputs to loop in the original graph but the outputs of the loop. It seems like a clearer fix is that this code should be manually cloning the loop node, inserting the right inputs, and only using the default cloning logic for the body of the loop. That way the right behavior is localized here and not changing the meeting of getOrAddInputForValue

Reply:
Let me try that! I actually thought that cloning a loop manually would be more work and more confusing in the end. We don't really have to change `getOrAddInputForValue` at all, I was trying to reuse the code at the expense of making that function more confusing and less coherent.